types/encode.go

```diff
@@ -63,35 +63,18 @@ func AtxAsBytes(tx *ActivationTx) ([]byte, error) {
 	return w.Bytes(), nil
 }
 
-func BytesAsAtx(b []byte) (*ActivationTx, error) {
+func BytesAsAtx(b []byte, id *AtxId) (*ActivationTx, error) {
 	buf := bytes.NewReader(b)
-	atx := ActivationTx{
-		ActivationTxHeader: ActivationTxHeader{
-			NIPSTChallenge: NIPSTChallenge{
-				NodeId: NodeId{
-					Key:          "",
-					VRFPublicKey: nil,
-				},
-				Sequence: 0,
-				PrevATXId: AtxId{
-					Hash: common.Hash{},
-				},
-				PubLayerIdx: 0,
-				StartTick:   0,
-				EndTick:     0,
-				PositioningAtx: AtxId{
-					Hash: common.Hash{},
-				},
-			},
-			ActiveSetSize: 0,
-			View:          nil,
-		},
-		Nipst: nil,
-	}
+	var atx ActivationTx
 	_, err := xdr.Unmarshal(buf, &atx)
 	if err != nil {
 		return nil, err
 	}
+	if id == nil {
```

Comment:
this does not belong here, this function only converts bytes to atx, set the ID outside to keep SOLID 'S' (Single Responsibility) for a function

Reply:
While I intuitively agree with you, the tradeoff here is that placing this code outside this function increases the risk of accidentally not adding the ID in some flows.

While I can check and verify that this is not the case _right now_, making it future-proof is harder.

It's worth noting that this functionality is not "hidden" inside this function, since the function takes the `id` param, revealing that something related is going on inside.

I'd be happy to implement a better solution if you can suggest something that's as safe and more SOLID.